File: ghidra/scripts/SweepCalibPointers.py
# Sweeps TriCore code region for LEA/MOVH+ADD pairs resolving to calibration region.
# Cross-references against known maps, validates data structures, decompiles callers.
# @category EDC17CP09
# @author Hermes

from ghidra.program.model.listing import *
from ghidra.program.model.address import *
from ghidra.program.model.mem import *
from ghidra.program.model.scalar import *
from ghidra.program.model.symbol import *
from ghidra.app.decompiler import *
from ghidra.program.model.pcode import *
from ghidra.util.task import TaskMonitor

import json
import os

CODE_START = 0x020000
CODE_END = 0x180000
CALIB_START = 0x040000
CALIB_END = 0x200000

program = currentProgram
listing = program.getListing()
memory = program.getMemory()
space = program.getAddressFactory().getDefaultAddressSpace()

# Load known map addresses from scan_mpc.json
knownAddresses = set()
try:
    jsonPath = "/home/brandon/Documents/EDC17CP09/scan_mpc.json"
    with open(jsonPath, 'r') as f:
        scan_data = json.load(f)
    for entry in scan_data.get('tables', []):
        addr = entry.get('address', 0)
        knownAddresses.add(addr)
except Exception as e:
    println("Warning: Could not load scan_mpc.json: %s" % str(e))

println("Known addresses from scan_mpc.json: %d" % len(knownAddresses))

# Sweep instructions in code region
codeStart = space.getAddress(CODE_START)
codeEnd = space.getAddress(CODE_END)

instrIter = listing.getInstructions(codeStart, True)
calibRefs = {}  # calib_addr -> [code_refs]

instrCount = 0
while instrIter.hasNext() and not monitor.isCancelled():
    instr = instrIter.next()
    instrAddr = instr.getAddress()
    addr = instrAddr.getOffset()
    if addr >= CODE_END:
        break
    instrCount += 1

    # Check each operand for references into calibration region
    for i in range(instr.getNumOperands()):
        try:
            ops = instr.getOpObjects(i)
            for op in ops:
                if isinstance(op, Scalar):
                    val = op.getUnsignedValue()
                    if CALIB_START <= val < CALIB_END:
                        if val not in calibRefs:
                            calibRefs[val] = []
                        calibRefs[val].append(addr)
                elif isinstance(op, Address):
                    val = op.getOffset()
                    if CALIB_START <= val < CALIB_END:
                        if val not in calibRefs:
                            calibRefs[val] = []
                        calibRefs[val].append(addr)
        except Exception as e:
            pass

println("Instructions scanned: %d" % instrCount)
println("Unique calibration addresses referenced: %d" % len(calibRefs))

# Filter known maps
newRefs = {}
for calibAddr, codeRefs in calibRefs.items():
    if calibAddr not in knownAddresses:
        newRefs[calibAddr] = codeRefs

println("New (not in scan_mpc.json): %d" % len(newRefs))

# Structural heuristic validation
def classifyStructure(data):
    """Classify the data structure type based on heuristics."""
    # Check for all zeros or all FF
    allZero = all(b == 0 for b in data[:32])
    allFF = all(b == 0xFF for b in data[:32])
    if allZero or allFF:
        return "EMPTY"
    
    # Check for 1D scalar (single byte or word)
    # If bytes 2-31 are all zero or constant, likely a scalar
    restConstant = len(set(data[2:32])) <= 2
    if restConstant and (data[0] != 0 or data[1] != 0):
        return "SCALAR_1D"
    
    # Check for DTC mask array (sequential 2-byte values, many zeros mixed in)
    zeroCount = sum(1 for i in range(0, 32, 2) if data[i] == 0 and data[i+1] == 0)
    if 8 <= zeroCount <= 14:
        return "DTC_MASK"
    
    # Check for 2D map header (axis count bytes followed by data)
    axisCount = data[0] & 0xFF
    if 2 <= axisCount <= 32:
        # Check if next bytes look like axis values (monotonically increasing)
        monotonic = True
        for i in range(1, min(axisCount, 8)):
            if (data[i] & 0xFF) <= (data[i-1] & 0xFF) and data[i] != 0:
                monotonic = False
                break
        if monotonic:
            return "2D_MAP_WITH_AXIS"
    
    # Check for curve (small axis count + values)
    if 2 <= axisCount <= 16:
        return "CURVE"
    
    return "UNKNOWN"

def inferDimensions(structType, data):
    """Infer dimensions based on structure type."""
    if structType == "SCALAR_1D":
        return "1x1 (2 bytes)"
    elif structType == "DTC_MASK":
        return "N x 2 bytes (mask array)"
    elif structType == "2D_MAP_WITH_AXIS":
        rows = data[0] & 0xFF
        cols = data[rows + 1] & 0xFF if (rows + 1) < len(data) else 1
        if cols < 2 or cols > 32:
            cols = 1
        return "%d x %d" % (rows, cols)
    elif structType == "CURVE":
        pts = data[0] & 0xFF
        return "%d x 1 (curve)" % pts
    else:
        return "unknown"

def decompileFunction(func):
    """Decompile a function and return C code."""
    if func is None:
        return ""
    try:
        decompiler = state.getTool().getService(DecompileInterface)
        if decompiler is None:
            return "[decompiler unavailable]"
        results = decompiler.decompileFunction(func, 60, monitor)
        if results is not None and results.decompileCompleted():
            return results.getDecompiledFunction().getC()
    except Exception as e:
        return "[decompile error: %s]" % str(e)
    return "[no decompiled output]"

def categorizeByContext(context, structType):
    """Categorize map based on decompiled context."""
    ctxLower = context.lower()
    if "ews" in ctxLower or "wegfahr" in ctxLower or "immo" in ctxLower:
        return "EWS/Immobilizer"
    if "dfes" in ctxLower or "dtc" in ctxLower or "error" in ctxLower:
        return "DTC/Error"
    if "torque" in ctxLower or "drehmoment" in ctxLower or "begrenzung" in ctxLower:
        return "Torque/Limiter"
    if "rail" in ctxLower or "druck" in ctxLower or "pressure" in ctxLower:
        return "RailPressure"
    if "boost" in ctxLower or "lade" in ctxLower or "turbo" in ctxLower:
        return "Boost/VNT"
    if "inject" in ctxLower or "einspritz" in ctxLower or "injcrv" in ctxLower:
        return "Injection"
    if "egr" in ctxLower or "dpf" in ctxLower or "regener" in ctxLower:
        return "Emissions"
    if "accped" in ctxLower or "fahrer" in ctxLower or "pedal" in ctxLower:
        return "DriversWish"
    if structType == "DTC_MASK":
        return "DTC/Error"
    if structType == "SCALAR_1D":
        return "Scalar/Config"
    return "Uncategorized"

def bytesToHex(data, length):
    """Convert bytes to hex string."""
    return " ".join("%02x" % (b & 0xFF) for b in data[:min(length, len(data))])

# Validate and process each discovered pointer
results = []
validated = 0

for calibAddr, codeRefs in newRefs.items():
    # Read data at calibration address
    data = bytearray(256)
    try:
        memory.getBytes(space.getAddress(calibAddr), data)
    except Exception as e:
        continue
    
    # Classify structure type
    structType = classifyStructure(data)
    if structType == "UNKNOWN" or structType == "EMPTY":
        continue
    
    # Determine inferred dimensions
    dimensions = inferDimensions(structType, data)
    
    # Find calling function and decompile
    decompiledContext = ""
    for codeRef in codeRefs:
        func = listing.getFunctionContaining(space.getAddress(codeRef))
        if func is not None:
            decompiledContext = decompileFunction(func)
            break
    
    # Categorize by context
    category = categorizeByContext(decompiledContext, structType)
    
    result = {
        "calib_addr": "0x%X" % calibAddr,
        "code_refs": ["0x%X" % ref for ref in codeRefs],
        "struct_type": structType,
        "dimensions": dimensions,
        "category": category,
        "decompiled_context": decompiledContext[:500] + "..." if len(decompiledContext) > 500 else decompiledContext,
        "first_bytes_hex": bytesToHex(data, 32)
    }
    
    results.append(result)
    validated += 1
    
    if validated >= 2000:
        println("Hit 2000 validated limit, stopping.")
        break

println("Validated new maps: %d" % validated)

# Write results to JSON
outputPath = "/home/brandon/Documents/EDC17CP09/new_maps_discovered.json"
with open(outputPath, 'w') as f:
    json.dump(results, f, indent=2)

println("Results written to: %s" % outputPath)

File: ghidra/src/SweepCalibPointers.java
// Sweeps TriCore code region for LEA/MOVH+ADD pairs resolving to calibration region.
// Cross-references against known maps, validates data structures, decompiles callers.
// @category EDC17CP09
// @author Hermes

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.*;
import ghidra.program.model.address.*;
import ghidra.program.model.mem.*;
import ghidra.program.model.scalar.*;
import ghidra.program.model.symbol.*;
import ghidra.app.decompiler.*;
import ghidra.program.model.pcode.*;
import ghidra.util.task.TaskMonitor;

import java.io.*;
import java.util.*;

public class SweepCalibPointers extends GhidraScript {

    static final long CODE_START = 0x020000L;
    static final long CODE_END   = 0x180000L;
    static final long CALIB_START = 0x040000L;
    static final long CALIB_END   = 0x200000L;

    @Override
    protected void run() throws Exception {
        Program program = currentProgram;
        Listing listing = program.getListing();
        Memory memory = program.getMemory();
        AddressSpace space = program.getAddressFactory().getDefaultAddressSpace();

        // Load known map addresses from scan_mpc.json
        Set<Long> knownAddresses = new HashSet<>();
        try {
            String jsonPath = "/home/brandon/Documents/EDC17CP09/scan_mpc.json";
            BufferedReader br = new BufferedReader(new FileReader(jsonPath));
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = br.readLine()) != null) sb.append(line);
            br.close();
            // Simple parse: find "address": 0xNNN patterns
            String json = sb.toString();
            int idx = 0;
            while ((idx = json.indexOf("\"address\"", idx)) >= 0) {
                int colon = json.indexOf(':', idx);
                if (colon < 0) break;
                int start = colon + 1;
                while (start < json.length() && !Character.isDigit(json.charAt(start)) && json.charAt(start) != '0') start++;
                int end = start;
                while (end < json.length() && (Character.isDigit(json.charAt(end)) || json.charAt(end) == 'x' || json.charAt(end) == 'X' || (json.charAt(end) >= 'a' && json.charAt(end) <= 'f') || (json.charAt(end) >= 'A' && json.charAt(end) <= 'F'))) end++;
                String numStr = json.substring(start, end).trim();
                try {
                    long addr;
                    if (numStr.startsWith("0x") || numStr.startsWith("0X")) {
                        addr = Long.parseLong(numStr.substring(2), 16);
                    } else {
                        addr = Long.parseLong(numStr);
                    }
                    knownAddresses.add(addr);
                } catch (NumberFormatException e) {}
                idx = end;
            }
        } catch (Exception e) {
            println("Warning: Could not load scan_mpc.json: " + e.getMessage());
        }
        println("Known addresses from scan_mpc.json: " + knownAddresses.size());

        // Sweep instructions in code region
        Address codeStart = space.getAddress(CODE_START);
        Address codeEnd = space.getAddress(CODE_END);

        InstructionIterator iter = listing.getInstructions(codeStart, true);
        Map<Long, List<Long>> calibRefs = new HashMap<>(); // calib_addr -> [code_refs]

        int instrCount = 0;
        while (iter.hasNext() && !monitor.isCancelled()) {
            Instruction instr = iter.next();
            Address instrAddr = instr.getAddress();
            long addr = instrAddr.getOffset();
            if (addr >= CODE_END) break;
            instrCount++;

            // Check each operand for references into calibration region
            for (int i = 0; i < instr.getNumOperands(); i++) {
                RefType refType = instr.getOperandRefType(i);
                Address[] refs = null;
                try {
                    refs = instr.getOpObjects(i) instanceof Address[] ? (Address[]) instr.getOpObjects(i) : null;
                } catch (Exception e) {}

                // Try scalar operands
                Object[] ops = instr.getOpObjects(i);
                for (Object op : ops) {
                    if (op instanceof Scalar) {
                        long val = ((Scalar) op).getUnsignedValue();
                        if (val >= CALIB_START && val < CALIB_END) {
                            calibRefs.computeIfAbsent(val, k -> new ArrayList<>()).add(addr);
                        }
                    } else if (op instanceof Address) {
                        long val = ((Address) op).getOffset();
                        if (val >= CALIB_START && val < CALIB_END) {
                            calibRefs.computeIfAbsent(val, k -> new ArrayList<>()).add(addr);
                        }
                    }
                }
            }
        }
        println("Instructions scanned: " + instrCount);
        println("Unique calibration addresses referenced: " + calibRefs.size());

        // Filter known maps
        Map<Long, List<Long>> newRefs = new HashMap<>();
        for (Map.Entry<Long, List<Long>> entry : calibRefs.entrySet()) {
            if (!knownAddresses.contains(entry.getKey())) {
                newRefs.put(entry.getKey(), entry.getValue());
            }
        }
        println("New (not in scan_mpc.json): " + newRefs.size());

        // Structural heuristic validation for each discovered pointer
        List<Map<String, Object>> results = new ArrayList<>();
        int validated = 0;

        for (Map.Entry<Long, List<Long>> entry : newRefs.entrySet()) {
            long calibAddr = entry.getKey();
            List<Long> codeRefs = entry.getValue();

            // Read data at calibration address
            byte[] data = new byte[256];
            try {
                memory.getBytes(space.getAddress(calibAddr), data);
            } catch (Exception e) {
                continue;
            }

            // Classify structure type
            String structType = classifyStructure(data);
            if (structType.equals("UNKNOWN")) continue;

            // Determine inferred dimensions
            String dimensions = inferDimensions(structType, data);

            // Find calling function and decompile
            String decompiledContext = "";
            for (long codeRef : codeRefs) {
                Function func = listing.getFunctionContaining(space.getAddress(codeRef));
                if (func != null) {
                    decompiledContext = decompileFunction(func);
                    break;
                }
            }

            // Categorize by context
            String category = categorizeByContext(decompiledContext, structType);

            Map<String, Object> result = new HashMap<>();
            result.put("calib_addr", "0x" + Long.toHexString(calibAddr));
            result.put("code_refs", codeRefs.stream().map(a -> "0x" + Long.toHexString(a)).collect(java.util.stream.Collectors.toList()));
            result.put("struct_type", structType);
            result.put("dimensions", dimensions);
            result.put("category", category);
            result.put("decompiled_context", decompiledContext.length() > 500 ? decompiledContext.substring(0, 500) + "..." : decompiledContext);
            result.put("first_bytes_hex", bytesToHex(data, 32));

            results.add(result);
            validated++;

            if (validated >= 2000) {
                println("Hit 2000 validated limit, stopping.");
                break;
            }
        }

        println("Validated new maps: " + validated);

        // Write results to JSON
        String outputPath = "/home/brandon/Documents/EDC17CP09/new_maps_discovered.json";
        FileWriter fw = new FileWriter(outputPath);
        // Manual JSON construction for compatibility
        fw.write("[\n");
        for (int i = 0; i < results.size(); i++) {
            Map<String, Object> r = results.get(i);
            fw.write("  {\n");
            fw.write("    \"calib_addr\": \"" + r.get("calib_addr") + "\",\n");
            fw.write("    \"code_refs\": " + r.get("code_refs") + ",\n");
            fw.write("    \"struct_type\": \"" + r.get("struct_type") + "\",\n");
            fw.write("    \"dimensions\": \"" + r.get("dimensions") + "\",\n");
            fw.write("    \"category\": \"" + r.get("category") + "\",\n");
            fw.write("    \"first_bytes_hex\": \"" + r.get("first_bytes_hex") + "\",\n");
            String ctx = ((String) r.get("decompiled_context")).replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n");
            fw.write("    \"decompiled_context\": \"" + ctx + "\"\n");
            fw.write("  }" + (i < results.size() - 1 ? "," : "") + "\n");
        }
        fw.write("]\n");
        fw.close();
        println("Results written to: " + outputPath);
    }

    String classifyStructure(byte[] data) {
        // Check for all zeros or all FF
        boolean allZero = true, allFF = true;
        for (int i = 0; i < 32; i++) {
            if (data[i] != 0) allZero = false;
            if (data[i] != (byte) 0xFF) allFF = false;
        }
        if (allZero || allFF) return "EMPTY";

        // Check for 1D scalar (single byte or word)
        // If bytes 2-31 are all zero or constant, likely a scalar
        boolean restConstant = true;
        for (int i = 2; i < 32; i++) {
            if (data[i] != data[2]) { restConstant = false; break; }
        }
        if (restConstant && (data[0] != 0 || data[1] != 0)) return "SCALAR_1D";

        // Check for DTC mask array (sequential 2-byte values, many zeros mixed in)
        int zeroCount = 0;
        for (int i = 0; i < 32; i += 2) {
            if (data[i] == 0 && data[i+1] == 0) zeroCount++;
        }
        if (zeroCount >= 8 && zeroCount <= 14) return "DTC_MASK";

        // Check for 2D map header (axis count bytes followed by data)
        int axisCount = data[0] & 0xFF;
        if (axisCount >= 2 && axisCount <= 32) {
            // Check if next bytes look like axis values (monotonically increasing)
            boolean monotonic = true;
            for (int i = 1; i < Math.min(axisCount, 8); i++) {
                if ((data[i] & 0xFF) <= (data[i-1] & 0xFF) && data[i] != 0) {
                    monotonic = false;
                    break;
                }
            }
            if (monotonic) return "2D_MAP_WITH_AXIS";
        }

        // Check for curve (small axis count + values)
        if (axisCount >= 2 && axisCount <= 16) return "CURVE";

        return "UNKNOWN";
    }

    String inferDimensions(String structType, byte[] data) {
        switch (structType) {
            case "SCALAR_1D":
                return "1x1 (2 bytes)";
            case "DTC_MASK":
                return "N x 2 bytes (mask array)";
            case "2D_MAP_WITH_AXIS":
                int rows = data[0] & 0xFF;
                int cols = data[rows + 1] & 0xFF;
                if (cols < 2 || cols > 32) cols = 1;
                return rows + " x " + cols;
            case "CURVE":
                int pts = data[0] & 0xFF;
                return pts + " x 1 (curve)";
            default:
                return "unknown";
        }
    }

    String decompileFunction(Function func) {
        if (func == null) return "";
        try {
            DecompileResults results;
            DecompileInterface decompiler = state.getTool().getService(DecompileInterface.class);
            if (decompiler == null) return "[decompiler unavailable]";
            results = decompiler.decompileFunction(func, 60, monitor);
            if (results != null && results.decompileCompleted()) {
                return results.getDecompiledFunction().getC();
            }
        } catch (Exception e) {
            return "[decompile error: " + e.getMessage() + "]";
        }
        return "[no decompiled output]";
    }

    String categorizeByContext(String context, String structType) {
        String ctxLower = context.toLowerCase();
        if (ctxLower.contains("ews") || ctxLower.contains("wegfahr") || ctxLower.contains("immo")) return "EWS/Immobilizer";
        if (ctxLower.contains("dfes") || ctxLower.contains("dtc") || ctxLower.contains("error")) return "DTC/Error";
        if (ctxLower.contains("torque") || ctxLower.contains("drehmoment") || ctxLower.contains("begrenzung")) return "Torque/Limiter";
        if (ctxLower.contains("rail") || ctxLower.contains("druck") || ctxLower.contains("pressure")) return "RailPressure";
        if (ctxLower.contains("boost") || ctxLower.contains("lade") || ctxLower.contains("turbo")) return "Boost/VNT";
        if (ctxLower.contains("inject") || ctxLower.contains("einspritz") || ctxLower.contains("injcrv")) return "Injection";
        if (ctxLower.contains("egr") || ctxLower.contains("dpf") || ctxLower.contains("regener")) return "Emissions";
        if (ctxLower.contains("accped") || ctxLower.contains("fahrer") || ctxLower.contains("pedal")) return "DriversWish";
        if (structType.equals("DTC_MASK")) return "DTC/Error";
        if (structType.equals("SCALAR_1D")) return "Scalar/Config";
        return "Uncategorized";
    }

    String bytesToHex(byte[] data, int len) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < Math.min(len, data.length); i++) {
            sb.append(String.format("%02x", data[i] & 0xFF));
            if (i < Math.min(len, data.length) - 1) sb.append(" ");
        }
        return sb.toString();
    }
}
